ParseException: Unparseable number: in any spinner input field

What were the actions performed prior to this error appearing?

1. Create any step with a number input box, (for example resize)
2. Enter a character in the spinner box.
3. Click the up or down buttons on the same spinner
4. Exception happens that can not be cleared.

Message: Thread[JavaFX Application Thread,5,main] threw an exception

## System Info:

Property Name | Property 
 ----- | -----
GRIP Version | 1.0.0-rc1
javafx.version | 8.0.66
java.runtime.name | Java(TM) SE Runtime Environment
java.vm.version | 25.66-b17
java.vm.vendor | Oracle Corporation
java.vm.name | Java HotSpot(TM) 64-Bit Server VM
java.runtime.version | 1.8.0_66-b17
java.awt.graphicsenv | sun.awt.CGraphicsEnvironment
javafx.runtime.version | 8.0.66-b17
os.name | Mac OS X
os.version | 10.11.2
os.arch | x86_64
file.encoding | UTF-8
java.vm.info | mixed mode
java.version | 1.8.0_66
sun.arch.data.model | 64
sun.cpu.endian | little

## Stack Trace:
```java
java.lang.RuntimeException: java.text.ParseException: Unparseable number: "d"
	at javafx.scene.control.SpinnerValueFactory$DoubleSpinnerValueFactory$1.fromString(SpinnerValueFactory.java:724)
	at javafx.scene.control.SpinnerValueFactory$DoubleSpinnerValueFactory$1.fromString(SpinnerValueFactory.java:696)
	at edu.wpi.grip.ui.pipeline.input.NumberSpinnerInputSocketController.commitEditorText(NumberSpinnerInputSocketController.java:91)
	at edu.wpi.grip.ui.pipeline.input.NumberSpinnerInputSocketController.lambda$initialize$79(NumberSpinnerInputSocketController.java:62)
	at com.sun.javafx.binding.ExpressionHelper$Generic.fireValueChangedEvent(ExpressionHelper.java:361)
	at com.sun.javafx.binding.ExpressionHelper.fireValueChangedEvent(ExpressionHelper.java:81)
	at javafx.beans.property.ReadOnlyBooleanPropertyBase.fireValueChangedEvent(ReadOnlyBooleanPropertyBase.java:72)
	at javafx.scene.Node$FocusedProperty.notifyListeners(Node.java:7718)
	at javafx.scene.Node.setFocused(Node.java:7771)
	at javafx.scene.Scene$KeyHandler.setWindowFocused(Scene.java:3932)
	at javafx.scene.Scene$KeyHandler.lambda$new$11(Scene.java:3954)
	at com.sun.javafx.binding.ExpressionHelper$SingleInvalidation.fireValueChangedEvent(ExpressionHelper.java:137)
	at com.sun.javafx.binding.ExpressionHelper.fireValueChangedEvent(ExpressionHelper.java:81)
	at javafx.beans.property.ReadOnlyBooleanWrapper$ReadOnlyPropertyImpl.fireValueChangedEvent(ReadOnlyBooleanWrapper.java:178)
	at javafx.beans.property.ReadOnlyBooleanWrapper$ReadOnlyPropertyImpl.access$100(ReadOnlyBooleanWrapper.java:148)
	at javafx.beans.property.ReadOnlyBooleanWrapper.fireValueChangedEvent(ReadOnlyBooleanWrapper.java:144)
	at javafx.beans.property.BooleanPropertyBase.markInvalid(BooleanPropertyBase.java:110)
	at javafx.beans.property.BooleanPropertyBase.set(BooleanPropertyBase.java:144)
	at javafx.stage.Window.setFocused(Window.java:439)
	at com.sun.javafx.stage.WindowPeerListener.changedFocused(WindowPeerListener.java:59)
	at com.sun.javafx.tk.quantum.GlassWindowEventHandler.run(GlassWindowEventHandler.java:100)
	at com.sun.javafx.tk.quantum.GlassWindowEventHandler.run(GlassWindowEventHandler.java:40)
	at java.security.AccessController.doPrivileged(Native Method)
	at com.sun.javafx.tk.quantum.GlassWindowEventHandler.lambda$handleWindowEvent$423(GlassWindowEventHandler.java:150)
	at com.sun.javafx.tk.quantum.QuantumToolkit.runWithoutRenderLock(QuantumToolkit.java:389)
	at com.sun.javafx.tk.quantum.GlassWindowEventHandler.handleWindowEvent(GlassWindowEventHandler.java:148)
	at com.sun.glass.ui.Window.handleWindowEvent(Window.java:1266)
	at com.sun.glass.ui.Window.notifyFocus(Window.java:1245)
	at com.sun.glass.ui.mac.MacWindow._setVisible(Native Method)
	at com.sun.glass.ui.Window.setVisible(Window.java:625)
	at com.sun.javafx.tk.quantum.WindowStage.lambda$setVisible$390(WindowStage.java:499)
	at com.sun.javafx.tk.quantum.QuantumToolkit.runWithRenderLock(QuantumToolkit.java:407)
	at com.sun.javafx.tk.quantum.WindowStage.setVisible(WindowStage.java:495)
	at javafx.stage.Window$9.invalidated(Window.java:869)
	at javafx.beans.property.BooleanPropertyBase.markInvalid(BooleanPropertyBase.java:109)
	at javafx.beans.property.BooleanPropertyBase.set(BooleanPropertyBase.java:144)
	at javafx.stage.Window.setShowing(Window.java:922)
	at javafx.stage.Window.show(Window.java:937)
	at javafx.stage.Stage.show(Stage.java:259)
	at javafx.stage.Stage.showAndWait(Stage.java:472)
	at javafx.scene.control.HeavyweightDialog.showAndWait(HeavyweightDialog.java:162)
	at javafx.scene.control.Dialog.showAndWait(Dialog.java:341)
	at edu.wpi.grip.ui.Main.lambda$onUnexpectedThrowableEvent$39(Main.java:83)
	at com.sun.javafx.application.PlatformImpl.runAndWait(PlatformImpl.java:317)
	at com.sun.javafx.application.PlatformImpl.runAndWait(PlatformImpl.java:307)
	at edu.wpi.grip.ui.Main.onUnexpectedThrowableEvent(Main.java:76)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at com.google.common.eventbus.EventSubscriber.handleEvent(EventSubscriber.java:74)
	at com.google.common.eventbus.SynchronizedEventSubscriber.handleEvent(SynchronizedEventSubscriber.java:47)
	at com.google.common.eventbus.EventBus.dispatch(EventBus.java:322)
	at com.google.common.eventbus.EventBus.dispatchQueuedEvents(EventBus.java:304)
	at com.google.common.eventbus.EventBus.post(EventBus.java:275)
	at edu.wpi.grip.core.GRIPCoreModule.onThreadException(GRIPCoreModule.java:124)
	at java.lang.ThreadGroup.uncaughtException(ThreadGroup.java:1057)
	at java.lang.ThreadGroup.uncaughtException(ThreadGroup.java:1052)
	at com.sun.glass.ui.Application.reportException(Application.java:453)
Caused by: java.text.ParseException: Unparseable number: "d"
	at java.text.NumberFormat.parse(NumberFormat.java:385)
	at javafx.scene.control.SpinnerValueFactory$DoubleSpinnerValueFactory$1.fromString(SpinnerValueFactory.java:722)
	... 58 more

```

--- a/core/src/main/java/edu/wpi/grip/core/operations/networktables/NTManager.java
+++ b/core/src/main/java/edu/wpi/grip/core/operations/networktables/NTManager.java
@@ -60,7 +60,7 @@ public class NTManager {
 
         NetworkTable.shutdown();
         if (projectSettings.getNetworkProtocol() == ProjectSettings.NetworkProtocol.NETWORK_TABLES) {
-            NetworkTable.setIPAddress(projectSettings.computeNetworkTablesServer());
+            NetworkTable.setIPAddress(projectSettings.computeNetworkProtocolServerAddress());
             NetworkTable.initialize();
         }
     }

--- a/core/src/main/java/edu/wpi/grip/core/settings/ProjectSettings.java
+++ b/core/src/main/java/edu/wpi/grip/core/settings/ProjectSettings.java
@@ -27,9 +27,13 @@ public class ProjectSettings implements Cloneable {
     @Setting(label = "FRC Team Number", description = "The team number, if used for FRC")
     private int teamNumber = 0;
 
-    @Setting(label = "NetworkTables Server", description = "The host that runs the NetworkTables server.  If not " +
-            "specified but NetworkTables is enabled, the hostname is derived from the team number.")
-    private String networkTablesServer = "";
+    @Setting(label = "NetworkTables Server Address", description = "The host that runs the Network Protocol server.  If not " +
+            "specified and Network Tables is specified as the protocol, the hostname is derived from the team number.")
+    private String networkProtocolServerAddress = "";
+
+    @Setting(label = "Deploy Address", description = "The remote host that grip should be remotely deployed to. If not " +
+            "specified and Network Tables is specified as the protocol, the hostname is derived from the team number.")
+    private String deployAddress = "";
 
     public void setNetworkProtocol(NetworkProtocol networkProtocol) {
         this.networkProtocol = checkNotNull(networkProtocol, "Network protocol cannot be null");
@@ -48,24 +52,40 @@ public class ProjectSettings implements Cloneable {
         return teamNumber;
     }
 
-    public void setNetworkTablesServer(String networkTablesServer) {
-        this.networkTablesServer =
-                checkNotNull(networkTablesServer, "NetworkTables server cannot be null");
+    public void setNetworkProtocolServerAddress(String networkProtocolServerAddress) {
+        this.networkProtocolServerAddress =
+                checkNotNull(networkProtocolServerAddress, "Network Protocol Server Address cannot be null");
     }
 
-    public String getNetworkTablesServer() {
-        return networkTablesServer;
+    public String getNetworkProtocolServerAddress() {
+        return networkProtocolServerAddress;
+    }
+
+    public void setDeployAddress(String deployAddress) {
+        this.deployAddress = checkNotNull(deployAddress, "Deploy Address can not be null");
+    }
+
+    public String getDeployAddress() {
+        return deployAddress;
     }
 
     /**
      * @return The address of the machine that the NetworkTables server is running on.  If
-     * {@link #setNetworkTablesServer} is specified, that is returned, otherwise this is based on the team number.
+     * {@link #setNetworkProtocolServerAddress} is specified, that is returned, otherwise this is based on the team number.
      */
-    public String computeNetworkTablesServer() {
-        if (networkTablesServer.isEmpty()) {
+    public String computeNetworkProtocolServerAddress() {
+        if (networkProtocolServerAddress.isEmpty()) {
             return "roboio-" + teamNumber + "-frc.local";
         } else {
-            return networkTablesServer;
+            return networkProtocolServerAddress;
+        }
+    }
+
+    public String computeDeployAddresss() {
+        if (networkProtocolServerAddress.isEmpty()) {
+            return "roboio-" + teamNumber + "-frc.local";
+        } else {
+            return networkProtocolServerAddress;
         }
     }
 
@@ -73,7 +93,7 @@ public class ProjectSettings implements Cloneable {
     public String toString() {
         return MoreObjects.toStringHelper(this)
                 .add("networkProtocol", networkProtocol)
-                .add("networkTablesServer", networkTablesServer)
+                .add("networkProtocolServerAddress", networkProtocolServerAddress)
                 .add("teamNumber", teamNumber)
                 .toString();
     }

--- a/ui/src/main/java/edu/wpi/grip/ui/DeployerController.java
+++ b/ui/src/main/java/edu/wpi/grip/ui/DeployerController.java
@@ -12,11 +12,9 @@ import edu.wpi.grip.ui.deployment.DeploymentOptionsControllersFactory;
 import edu.wpi.grip.ui.util.deployment.DeployedInstanceManager;
 import javafx.application.Platform;
 import javafx.fxml.FXML;
-import javafx.scene.control.Accordion;
-import javafx.scene.control.DialogPane;
-import javafx.scene.control.ProgressBar;
-import javafx.scene.control.TextArea;
+import javafx.scene.control.*;
 import javafx.scene.layout.HBox;
+import javafx.scene.layout.Priority;
 
 import java.io.IOException;
 import java.io.OutputStream;
@@ -122,9 +120,17 @@ public class DeployerController {
                 })
                 .done(deployedManager -> {
                     Platform.runLater(() -> {
-                        controlsBox.getChildren().add(startStopButtonFactory.create(deployedManager));
+                        final Label startStopLabel = new Label("Start / Stop Deployed Instance: ");
+                        startStopLabel.setMaxWidth(Double.MAX_VALUE);
+                        startStopLabel.setMaxHeight(Double.MAX_VALUE);
+                        HBox.setHgrow(startStopLabel, Priority.SOMETIMES);
+                        final StartStoppableButton startStoppableButton = startStopButtonFactory.create(deployedManager);
+                        startStopLabel.setLabelFor(startStoppableButton);
+                        controlsBox.getChildren().addAll(startStopLabel, startStoppableButton);
                         deploymentMethods.setDisable(true);
                         progressIndicator.setProgress(-1);
+                        // Resize window to accommodate the added label
+                        getRoot().getScene().getWindow().sizeToScene();
                     });
                 });
 

--- a/ui/src/main/java/edu/wpi/grip/ui/deployment/DeploymentOptionsController.java
+++ b/ui/src/main/java/edu/wpi/grip/ui/deployment/DeploymentOptionsController.java
@@ -77,6 +77,7 @@ public abstract class DeploymentOptionsController {
                     });
                 })
                 .done((t) -> {
+                    setErrorText("Success! Waiting for start.");
                     onDeployCallback.accept(t);
                     Platform.runLater(() -> {
                         deploySpinner.setVisible(false);
@@ -89,6 +90,7 @@ public abstract class DeploymentOptionsController {
         Platform.runLater(() -> {
             deployErrorText.setText(text);
             root.requestLayout();
+            root.getScene().getWindow().sizeToScene();
         });
     }
 
@@ -121,7 +123,8 @@ public abstract class DeploymentOptionsController {
                     if (inetAddress.isReachable(1000)) {
                         return inetAddress;
                     } else {
-                        notify("Attempt " + i + "/" + attemptCount + " failed");
+                        // i + 1 so that the attempt counter is zero based.
+                        notify("Attempt " + i + 1 + "/" + attemptCount + " failed");
                     }
                 }
                 throw new IOException("Failed to connect");

--- a/ui/src/main/java/edu/wpi/grip/ui/deployment/FRCAdvancedDeploymentOptionsController.java
+++ b/ui/src/main/java/edu/wpi/grip/ui/deployment/FRCAdvancedDeploymentOptionsController.java
@@ -2,6 +2,7 @@ package edu.wpi.grip.ui.deployment;
 
 import com.google.inject.Inject;
 import com.google.inject.assistedinject.Assisted;
+import edu.wpi.grip.core.Pipeline;
 import edu.wpi.grip.ui.util.deployment.DeployedInstanceManager;
 import javafx.scene.control.Label;
 import javafx.scene.control.TextField;
@@ -21,6 +22,7 @@ public class FRCAdvancedDeploymentOptionsController extends DeploymentOptionsCon
     private final Supplier<OutputStream> stdOut, stdErr;
 
     private TextField address;
+    private final String projectAddress;
 
     public interface Factory {
         FRCAdvancedDeploymentOptionsController create(
@@ -32,11 +34,13 @@ public class FRCAdvancedDeploymentOptionsController extends DeploymentOptionsCon
 
     @Inject
     FRCAdvancedDeploymentOptionsController(DeployedInstanceManager.Factory deployedInstanceManagerFactor,
+                                           Pipeline pipeline,
                                            @Assisted Consumer<DeployedInstanceManager> onDeployCallback,
                                            @Assisted("stdOut") Supplier<OutputStream> stdOut,
                                            @Assisted("stdErr") Supplier<OutputStream> stdErr) {
         super("FRC Advanced", onDeployCallback);
         this.deployedInstanceManagerFactor = deployedInstanceManagerFactor;
+        this.projectAddress = pipeline.getProjectSettings().computeDeployAddresss();
         this.stdOut = stdOut;
         this.stdErr = stdErr;
 
@@ -45,6 +49,7 @@ public class FRCAdvancedDeploymentOptionsController extends DeploymentOptionsCon
     @Override
     protected void postInit() {
         this.address = new TextField();
+        this.address.setText(projectAddress);
         this.address.setPromptText("roborio-[team number]-frc.local");
 
         this.address.textProperty().addListener((observable, oldValue, newValue) -> {

--- a/ui/src/main/java/edu/wpi/grip/ui/deployment/FRCDeploymentOptionsController.java
+++ b/ui/src/main/java/edu/wpi/grip/ui/deployment/FRCDeploymentOptionsController.java
@@ -2,6 +2,8 @@ package edu.wpi.grip.ui.deployment;
 
 import com.google.inject.Inject;
 import com.google.inject.assistedinject.Assisted;
+import edu.wpi.grip.core.Pipeline;
+import edu.wpi.grip.ui.util.Spinners;
 import edu.wpi.grip.ui.util.deployment.DeployedInstanceManager;
 import javafx.scene.control.Label;
 import javafx.scene.control.Spinner;
@@ -20,6 +22,7 @@ import static com.google.common.base.Preconditions.checkNotNull;
 public class FRCDeploymentOptionsController extends DeploymentOptionsController {
 
     private Spinner<Integer> teamNumberSpinner;
+    private final int teamNumber;
     private final DeployedInstanceManager.Factory deployedInstanceManagerFactor;
     private final Supplier<OutputStream> stdOut, stdErr;
 
@@ -33,35 +36,30 @@ public class FRCDeploymentOptionsController extends DeploymentOptionsController 
 
     @Inject
     FRCDeploymentOptionsController(DeployedInstanceManager.Factory deployedInstanceManagerFactor,
+                                   Pipeline pipeline,
                                    @Assisted Consumer<DeployedInstanceManager> onDeployCallback,
                                    @Assisted("stdOut") Supplier<OutputStream> stdOut,
                                    @Assisted("stdErr") Supplier<OutputStream> stdErr) {
         super("FRC", onDeployCallback);
         this.deployedInstanceManagerFactor = deployedInstanceManagerFactor;
+        this.teamNumber = pipeline.getProjectSettings().getTeamNumber();
         this.stdOut = stdOut;
         this.stdErr = stdErr;
     }
 
     @Override
-    @SuppressWarnings("PMD.IfElseStmtsMustUseBraces")
     protected void postInit() {
         final Label label = new Label("Team Number");
-        final SpinnerValueFactory.IntegerSpinnerValueFactory spinnerValueFactory = new SpinnerValueFactory.IntegerSpinnerValueFactory(0, Integer.MAX_VALUE, 190);
+        final SpinnerValueFactory.IntegerSpinnerValueFactory spinnerValueFactory =
+                new SpinnerValueFactory.IntegerSpinnerValueFactory(0, Integer.MAX_VALUE, teamNumber);
         this.teamNumberSpinner = new Spinner(spinnerValueFactory);
-        this.teamNumberSpinner.setEditable(true);
-        // Ensure the value entered is only a number
-        this.teamNumberSpinner.getEditor().textProperty().addListener((observable, oldValue, newValue) -> {
-            if ("".equals(newValue)) {
-                teamNumberSpinner.getEditor().setText(Integer.toString(0));
-            } else try {
-                int value = Integer.parseInt(newValue);
-                teamNumberSpinner.getEditor().setText(Integer.toString(value));
-            } catch (NumberFormatException e) {
-                teamNumberSpinner.getEditor().setText(oldValue);
-            }
-        });
+        Spinners.makeEditableSafely(teamNumberSpinner, Integer::valueOf);
+        label.setLabelFor(teamNumberSpinner);
+
         getOptionsGrid().addRow(0, label, this.teamNumberSpinner);
     }
+
+
 
     @Override
     protected Promise<DeployedInstanceManager, String, String> onDeploy() {

--- a/ui/src/main/java/edu/wpi/grip/ui/pipeline/input/NumberSpinnerInputSocketController.java
+++ b/ui/src/main/java/edu/wpi/grip/ui/pipeline/input/NumberSpinnerInputSocketController.java
@@ -7,11 +7,11 @@ import edu.wpi.grip.core.InputSocket;
 import edu.wpi.grip.core.events.SocketChangedEvent;
 import edu.wpi.grip.ui.pipeline.SocketHandleView;
 import edu.wpi.grip.ui.util.GRIPPlatform;
+import edu.wpi.grip.ui.util.Spinners;
 import javafx.beans.InvalidationListener;
 import javafx.fxml.FXML;
 import javafx.scene.control.Spinner;
 import javafx.scene.control.SpinnerValueFactory;
-import javafx.util.StringConverter;
 
 import static com.google.common.base.Preconditions.checkArgument;
 
@@ -55,12 +55,9 @@ public class NumberSpinnerInputSocketController extends InputSocketController<Nu
     public void initialize() {
         super.initialize();
         final Spinner<Double> spinner = new Spinner<>(this.valueFactory);
-        spinner.setEditable(true);
+        Spinners.makeEditableSafely(spinner, Double::valueOf);
         spinner.disableProperty().bind(this.getHandle().connectedProperty());
-        spinner.focusedProperty().addListener((s, ov, nv) -> {// Code found at http://stackoverflow.com/questions/32340476/manually-typing-in-text-in-javafx-spinner-is-not-updating-the-value-unless-user
-            if (nv) return;
-            commitEditorText(spinner);
-        });
+
         this.setContent(spinner);
     }
 
@@ -80,17 +77,4 @@ public class NumberSpinnerInputSocketController extends InputSocketController<Nu
         }
     }
 
-    // Code found at http://stackoverflow.com/questions/32340476/manually-typing-in-text-in-javafx-spinner-is-not-updating-the-value-unless-user
-    private <T> void commitEditorText(Spinner<T> spinner) {
-        if (!spinner.isEditable()) return;
-        String text = spinner.getEditor().getText();
-        SpinnerValueFactory<T> valueFactory = spinner.getValueFactory();
-        if (valueFactory != null) {
-            StringConverter<T> converter = valueFactory.getConverter();
-            if (converter != null) {
-                T value = converter.fromString(text);
-                valueFactory.setValue(value);
-            }
-        }
-    }
 }

--- a/ui/src/main/java/edu/wpi/grip/ui/util/Spinners.java
+++ b/ui/src/main/java/edu/wpi/grip/ui/util/Spinners.java
@@ -1,0 +1,61 @@
+package edu.wpi.grip.ui.util;
+
+
+import javafx.scene.control.Spinner;
+import javafx.scene.control.SpinnerValueFactory;
+import javafx.util.StringConverter;
+
+import java.util.function.Function;
+
+/**
+ * Utility methods to set up spinners in a safe way.
+ */
+public final class Spinners {
+
+    private Spinners() {
+        /* no-op */
+    }
+
+    /**
+     * Makes the spinner editable.
+     * Ensures that the values will be committed when focus is lost.
+     * Ensures that the only values that the spinner can accept are parable into whatever number type they represent.
+     * @param <T> The number type of the spinner
+     */
+    @SuppressWarnings("PMD.IfElseStmtsMustUseBraces")
+    public static <T extends Number> void makeEditableSafely(Spinner<T> spinner, Function<String, T> parseToNumber) {
+        spinner.setEditable(true);
+        spinner.focusedProperty().addListener((s, ov, nv) -> {
+            if (nv) return;
+            commitEditorText(spinner);
+        });
+        // Ensure the value entered is only a number
+        spinner.getEditor().textProperty().addListener((observable, oldValue, newValue) -> {
+            if ("".equals(newValue)) {
+                spinner.getEditor().setText("0");
+            } else try {
+                T value = parseToNumber.apply(newValue);
+                spinner.getEditor().setText(value.toString());
+            } catch (NumberFormatException e) {
+                spinner.getEditor().setText(oldValue);
+            }
+        });
+    }
+
+    /**
+     *
+     * @see <a href="http://stackoverflow.com/questions/32340476/manually-typing-in-text-in-javafx-spinner-is-not-updating-the-value-unless-user">Stack Overflow Post</a>
+     */
+    private static <T> void commitEditorText(Spinner<T> spinner) {
+        if (!spinner.isEditable()) return;
+        String text = spinner.getEditor().getText();
+        SpinnerValueFactory<T> valueFactory = spinner.getValueFactory();
+        if (valueFactory != null) {
+            StringConverter<T> converter = valueFactory.getConverter();
+            if (converter != null) {
+                T value = converter.fromString(text);
+                valueFactory.setValue(value);
+            }
+        }
+    }
+}

